### ISSUE TITLE
Python 3.9.0b1: AssertionError in BaseParser.apply_trailers()

### ISSUE THREAD

**Issue by laloch** [xonsh/xonsh#3585]
This may be due to the ongoing deprecation of ast.Num, ast.Str, ast.Bytes, ast.NameConstant and ast.Ellipsis.

For instance `ast.Str()` now creates `ast.Constant`, so `isinstance(ast.Str(), ast.Str)` is not `True` anymore.

Applies to both LL(1) and PEG parsers.

## xonfig

<details>

```
$ xonfig
+------------------+----------------------+
| xonsh            | 0.9.18.dev8          |
| Git SHA          | 6546c2b9             |
| Commit Date      | May 12 15:53:08 2020 |
| Python           | 3.9.0                |
| PLY              | 3.11                 |
| have readline    | True                 |
| prompt toolkit   | 2.0.4                |
| shell type       | prompt_toolkit       |
| pygments         | 2.6.1                |
| on posix         | True                 |
| on linux         | True                 |
| distro           | unknown              |
| on darwin        | False                |
| on windows       | False                |
| on cygwin        | False                |
| on msys2         | False                |
| is superuser     | True                 |
| default encoding | utf-8                |
| xonsh encoding   | utf-8                |
| encoding errors  | surrogateescape      |
+------------------+----------------------+
```

</details>

### Traceback (if applicable)

<details>

```
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/xonsh/ptk_shell/shell.py", line 181, in _push
    code = self.execer.compile(src, mode="single", glbs=self.ctx, locs=None)
  File "/usr/local/lib/python3.9/site-packages/xonsh/execer.py", line 135, in compile
    tree = self.parse(input, ctx, mode=mode, filename=filename, transform=transform)
  File "/usr/local/lib/python3.9/site-packages/xonsh/execer.py", line 95, in parse
    tree, input = self._parse_ctx_free(input, mode=mode, filename=filename)
  File "/usr/local/lib/python3.9/site-packages/xonsh/execer.py", line 219, in _parse_ctx_free
    tree = self.parser.parse(
  File "/usr/local/lib/python3.9/site-packages/xonsh/parsers/base.py", line 543, in parse
    tree = self.parser.parse(input=s, lexer=self.lexer, debug=debug_level)
  File "/usr/local/lib/python3.9/site-packages/xonsh/ply/ply/yacc.py", line 335, in parse
    return self.parseopt_notrack(input, lexer, debug, tracking, tokenfunc)
  File "/usr/local/lib/python3.9/site-packages/xonsh/ply/ply/yacc.py", line 1122, in parseopt_notrack
    p.callable(pslice)
  File "/usr/local/lib/python3.9/site-packages/xonsh/parsers/base.py", line 2249, in p_atom_expr
    p[0] = self.apply_trailers(p[1], p[2])
  File "/usr/local/lib/python3.9/site-packages/xonsh/parsers/base.py", line 2243, in apply_trailers
    assert False
AssertionError
```

</details>

## Steps to Reproduce

```
$ foo=["bar"]
$ foo[0]
```

**Comment by laloch** [comment]
Potential fix in laloch/xonsh@a137803ed4a16fc63d8e8314443e31ae890989c3.

**Comment by jnoortheen** [comment]
the issue is caused by the change https://docs.python.org/3/whatsnew/3.9.html?highlight=simplified%20ast%20subscription#changes-in-the-python-api here.

**Comment by anki-code** [comment]
Fixed in master by #4101.


Tested in docker:
```python
docker run --rm -it python:3.9-slim /bin/bash -c "apt update && apt install -y git && pip install -U git+https://github.com/xonsh/xonsh.git && xonsh"
```
